File: protocol/load.go
package protocol

import (
	"fmt"
	"github.com/gozelle/_fs"
	"gopkg.in/yaml.v3"
)

func Load(path string) (config *Config, err error) {
	ok, err := _fs.Exist(path)
	if err != nil {
		return
	}
	if !ok {
		err = fmt.Errorf("path: %s not exist", path)
		return
	}

	content, err := _fs.Read(path)
	if err != nil {
		return
	}
	config = new(Config)
	err = yaml.Unmarshal(content, config)
	if err != nil {
		err = fmt.Errorf("unmarshal yml error: %s", err)
		return
	}
	return
}

File: apps/ship/src/protocol/load.ts
import { existsSync, readFileSync } from "node:fs";
import { parse } from "yaml";
import { type Config, configSchema } from "./schema.js";

export function loadConfig(path: string): Config {
  if (!existsSync(path)) {
    throw new Error(`${path} not exist`);
  }
  const content = readFileSync(path, "utf8");
  const result = configSchema.safeParse(parse(content));
  if (!result.success) {
    throw new Error(`unmarshal yml error: ${result.error.message}`);
  }
  return result.data;
}
